packages/config-helpers/src/types.ts

```diff
@@ -0,0 +1,29 @@
+/**
+ * @fileoverview Types for this package.
+ */
+
+import type { Linter } from "eslint";
+
+/**
+ * The type of array element in the `extends` property after flattened.
+ */
+export type SimpleExtendsElement = string | Linter.Config;
+
+/**
+ * The type of array element in the `extends` property before flattening.
+ */
+export type ExtendsElement = SimpleExtendsElement | Linter.Config[];
+
+/**
+ * Config with extends. Valid only inside of `defineConfig()`.
+ */
+export interface ConfigWithExtends extends Linter.Config {
+	extends?: ExtendsElement[];
+}
+
+/**
+ * Infinite array type.
+ */
+type InfiniteArray<T> = InfiniteArray<T>[];
```

Comment:
This doesn't seem to allow anything except empty arrays and empty arrays in empty arrays.

For example, `defineConfig({})` or `defineConfig([{}])`, which should be valid, produce error `Type '{}' is missing the following properties from type 'InfiniteArray<ConfigWithExtends>[]': length, pop, push, concat, and 29 more.`.

I _think_ this should be something like `type InfiniteArray<T> = T | InfiniteArray<T>[];`.